golang/cosmos/x/swingset/keeper/extension_snapshotter.go

```diff
@@ -0,0 +1,297 @@
+}
+
+// NewExtensionSnapshotter creates a new swingset ExtensionSnapshotter
+func NewExtensionSnapshotter(app *baseapp.BaseApp, swingStoreExportsHandler *SwingStoreExportsHandler, getSwingStoreExportDataShadowCopy func(ctx sdk.Context) []*vstoragetypes.DataEntry) *ExtensionSnapshotter {
+	return &ExtensionSnapshotter{
+		isConfigured:    func() bool { return app.SnapshotManager() != nil },
+		takeAppSnapshot: app.Snapshot,
+		newRestoreContext: func(height int64) sdk.Context {
+			return app.NewUncachedContext(false, tmproto.Header{Height: height})
+		},
+		logger:                            app.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "extension snapshotter"),
+		swingStoreExportsHandler:          swingStoreExportsHandler,
+		getSwingStoreExportDataShadowCopy: getSwingStoreExportDataShadowCopy,
+		activeSnapshot:                    nil,
+	}
+}
+
+// SnapshotName returns the name of the snapshotter, it should be unique in the manager.
+// Implements ExtensionSnapshotter
+func (snapshotter *ExtensionSnapshotter) SnapshotName() string {
+	return types.ModuleName
+}
+
+// SnapshotFormat returns the default format the extension snapshotter uses to encode the
+// payloads when taking a snapshot.
+// It's defined within the extension, different from the global format for the whole state-sync snapshot.
+// Implements ExtensionSnapshotter
+func (snapshotter *ExtensionSnapshotter) SnapshotFormat() uint32 {
+	return SnapshotFormat
+}
+
+// SupportedFormats returns a list of formats it can restore from.
+// Implements ExtensionSnapshotter
+func (snapshotter *ExtensionSnapshotter) SupportedFormats() []uint32 {
+	return []uint32{SnapshotFormat}
+}
+
+// InitiateSnapshot initiates a snapshot for the given block height.
+// If a snapshot is already in progress, or if no snapshot manager is
+// configured, this will fail.
+// The snapshot operation is performed in a goroutine managed by the
+// SwingStoreExportsHandler. Use WaitUntilSwingStoreExportStarted to
+// synchronize commit boundaries.
+func (snapshotter *ExtensionSnapshotter) InitiateSnapshot(height int64) error {
+	if !snapshotter.isConfigured() {
+		return fmt.Errorf("snapshot manager not configured")
+	} else if height <= 0 {
+		return fmt.Errorf("block height must not be negative or 0")
+	}
+
+	blockHeight := uint64(height)
+
+	return snapshotter.swingStoreExportsHandler.InitiateExport(blockHeight, snapshotter, SwingStoreExportOptions{
+		ExportMode:        "current",
+		IncludeExportData: false,
+	})
+}
+
+// ExportInitiated performs the actual cosmos state-sync app snapshot.
+// The cosmos implementation will ultimately call SnapshotExtension, which can
+// retrieve and process the SwingStore artifacts.
+// This method is invoked by the SwingStoreExportsHandler in a goroutine
+// started by InitiateExport, only if no other SwingStore export operation is
+// already in progress.
+//
+// Implements SwingStoreExportEventHandler
```

Comment:
Comment before the rebase:

```suggestion
// TakeSnapshot sends a bridge message to request that Swingset perform a SwingStore export and
// then uses retrieveSwingStoreExport to retrieve it.
// This function will block until the export has been retrieved.
// No other SwingStore export may be in progress.
// Implements SwingStoreSnapshotTaker.
```

Reply:
This suggestion is not accurate. I think this is rooted in the same confusion as @gibson042 regarding the inversion of responsibilities that the current layering suffers from. The rebase clarifies this, but in general I will follow up to clean up the layering and better assign responsibilities between the components, at which point I will also create a diagram to help visualize the flow.